### ISSUE TITLE
Question On Schema Veritication

### ISSUE THREAD

**Issue by SoundsSerious** [threatify/arango-orm#97]
Definetly Loving Arango ORM! Makes it even easier to use arangodb!

One use case that I'm hoping to understand better is the schema verification. I see in #17 that there was a discussion of a first class schema definition.

I'm no Arango expert but looking at pyarango docs it seems officially a schema field is supported, however it doesn't appear that arango-orm utilizes on DB schema? Is this correct?

As far as motivation: My goal is to use arango-orm or python equivalent to define some backbone of an API that other programmers might use (not always python) and I'd hope to have some type of protections that are universal across the entire api.

What would be the best way to enforce a standard defined in arango-orm across all connected applications?

**Comment by kashifpk** [comment]
The current method of using marshmallow was developed before arango introduced support for json schema. There is a discussion about that here https://github.com/threatify/arango-orm/issues/32

While using the built-in json schema gives us validation, marshmallow offers serialization capabilities too. This is one of the reasons for still using marshmallow.

For example you could fetch data from arango and serve it up via an API. The API response could call Collection._dump() method to make a dict that is json serializable. It is just more convenient. That said, we haven't totally ruled out using jsonschema as that looks like it's going to be a standard and it can be applied directly on in-database data instead of doing the validations on data insertion or loading that is currently done by arango-orm.

> What would be the best way to enforce a standard defined in arango-orm across all connected applications?

If you describe the validations in model class definitions then at API level or in other python applications you can import those models. When using arango-orm this would ensure that you have consistent data. If other non-python applications have only access to your data via your API then this means that your validation is automatically applied everywhere.

**Comment by SoundsSerious** [comment]
Hi @kashifpk thanks for the detailed response.

I definitely agree its working just find as is, however I would definitely love to have a way to take advantage of defining a JSON schema even if it is just a pass through. My concern is that I would like to define a schema that others will use with other languages so unfortunately the in-python validation wont work for javascript / node ect.

It does seem like there's a relatively easy solution via this `python-arango` example: https://docs.python-arango.com/en/main/schema.html

It seems the underlying driver already has support for schema, however we just need a way to pass this through to the collection. In a short examination of the ORM code I didn't see where I might add a schema definition in creation, however i could still use the underlying driver to modify the collection.  Not sure of all the particulars but it seems there could be a good solution that automatically converts marshmellow requirements into these schema-dicts. 

```
# Create a new collection named "employees" with custom schema.
my_schema = {
    'rule': {
        'type': 'object',
        'properties': {
            'name': {'type': 'string'},
            'email': {'type': 'string'}
        },
        'required': ['name', 'email']
    },
    'level': 'moderate',
    'message': 'Schema Validation Failed.'
}
employees = db.create_collection(name='employees', schema=my_schema)

# Modify the schema.
employees.configure(schema=my_schema)

# Remove the schema.
employees.configure(schema={})
```

**Comment by SoundsSerious** [comment]
What would be the recommended way, if any for adding this first-class db schema validation?

**Comment by SoundsSerious** [comment]
I'm thinking the only possible way to do this, while using `arango-orm` would be to make a embed a function that calls `python-arango` and adds this schema dictionary.

Is there another option that i'm not thinking of?

**Comment by kashifpk** [comment]
I was thinking along the lines of a  method to convert between jsonschema and marshmallow class definition so there is no code duplication. Basically we could have a switch within the model classes like `_store_schema = True` so the fields and constraints that we have defined in arango-orm are converted to jsonschema representation and stored with the collection when we create the collection via arango-orm.

Similarly support for dynamic/anonymous models is underway where you can fetch documents from any collection and you have not created model classes for those collections. Currently this just creates a class dynamically and loads the data so you can access it in object notation just like normal model classes. Of course these classes don't have any validation or extra functionality since the are created dynamically. I'm planning to make use of jsonschema so if the collection has a schema the constraints defined there are loaded and we get a dynamic model class that also has jsonschema constraints enabled.

**Comment by SoundsSerious** [comment]
I like that idea! I definitely see that as a solution! 

I can imagine though it'll take some time flesh out all the use cases so as not to break existing apps. Would be nice to just be able to flip the `store_schema` flag and have it all work in DB :)

I think maybe intermediate / complementary solution would be to expose a field during collection creation. I think the following would just rely on a dictionary passthrough  to the driver. Later if you want to change it thats on you (maybe you get a warning if schema is invalid or something). 

`db.create_collection(Student, db_schema =  {manual_entry_stuff} )` 

Maybe if a private variable is defined in class like `__db_schema` this is automatically done, and can be a placeholder for later.

I think that would lay the ground work for more development on first class schema validation!

**Comment by kashifpk** [comment]
Yes for create_collection we would definitely need a parameter. Later it can be overridden or determined automatically based on collection structure. Adding this to things to do for next release.
